### PR TITLE
[InstallAPI] Pick up input headers by directory traversal

### PR DESCRIPTION
**Pull request by cyndyishida**
Match TAPI behavior and allow input headers to be resolved via a passed directory, which is expected to be a library sitting in a build directory.

### CODE REVIEW

File: clang/include/clang/InstallAPI/HeaderFile.h [hunk 1]
```diff
@@ -97,6 +97,14 @@ class HeaderFile {
                                           Other.Excluded, Other.Extra,
                                           Other.Umbrella);
   }
+
+  bool operator<(const HeaderFile &Other) const {
```
Comment (zixu-w) [inline clang/include/clang/InstallAPI/HeaderFile.h:101]: What's the reasoning for the ordering? Might be easier to understand in the future with some comments.

Comment (cyndyishida) [reply clang/include/clang/InstallAPI/HeaderFile.h:101]: I could infer, but I don't actually know @ributzka do you remember?

Comment (ributzka) [reply clang/include/clang/InstallAPI/HeaderFile.h:101]: Including the umbrella header first increases the likelihood of successfully parsing all the headers in the framework. Extra headers always go last.

**Comment by cyndyishida** [comment]
ping

**Review by zixu-w — APPROVED** [review]

File: clang/lib/InstallAPI/DirectoryScanner.cpp [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+  }
+
+  // Go through the subdirectories.
+  // Sort the sub-directory first since different file systems might have
+  // different traverse order.
+  llvm::sort(SubDirectories);
+  if (ParentPath.empty())
+    ParentPath = Path;
+  for (const StringRef Dir : SubDirectories)
+    return scanHeaders(Dir, Lib, Type, BasePath, ParentPath);
```
[the remark this answers is not in the document]
Comment (cyndyishida) [reply clang/lib/InstallAPI/DirectoryScanner.cpp:133]: Thanks for reporting! fixed in: https://github.com/llvm/llvm-project/pull/100636

**Pull request merged**